src/analyzer/protocol/quic/QUIC.spicy

```diff
@@ -453,50 +530,55 @@ type Packet = unit(from_client: bool, context: ConnectionIDInfo&) {
         # If we see a retry packet from the responder, reset the decryption
         # context such that the next DCID from the client is used for decryption.
         if ( self.long_header.is_retry ) {
-          context.client_initial_processed = False;
-          context.server_initial_processed = False;
-          context.initial_destination_conn_id = b"";
 
           # Allow re-opening the SSL analyzer the next time around.
           zeek::protocol_handle_close(context.ssl_handle);
           unset context.ssl_handle;
+
+          # Reset crypto state!
+          context.initial_destination_conn_id = b"";
+
+          context.client_crypto.flushed = False;
+          crypto_state_connect(context.client_crypto);
+
+          context.server_crypto.flushed = False;
+          crypto_state_connect(context.server_crypto);
         }
      }
   };
 
   : void {
-      if (self?.long_header && can_decrypt(self.long_header, context, from_client))
+      if ( self?.long_header && can_decrypt(self.long_header, context, self.crypto_state) )
           # If we have parsed an initial packet that we can decrypt the payload,
           # determine the size to store into a buffer.
           self.packet_size = self.offset();
   }
 
   # Buffer the whole packet if we determined we have a chance to decrypt.
   packet_data: bytes &parse-at=self.start &size=self.packet_size if ( self.packet_size > 0 ) {
-    self.crypto_buffer = new CryptoBuffer();
-    self.crypto_sink.connect(self.crypto_buffer);
 
     if ( from_client ) {
       context.server_cid_len = self.long_header.dest_conn_id_len;
       context.client_cid_len = self.long_header.src_conn_id_len;
 
+      # This is the first INITIAL packet we attempt to decrypt and it is
+      # coming from the client. Use its destination connection ID for
+      # decryption purposes.
+      if ( |context.initial_destination_conn_id| == 0 ) {
```

Comment:
Nit: `initial_destination_conn_id` should really be an `optional<bytes>`, would you mind cleaning that up? With that the length checks would become just `!context.initial_destination_conn_id`, the "reset" above would assign `Null` instead, and value access would deref.